src/lib/fw_performance_model/PerformanceModel.cpp

```diff
@@ -0,0 +1,238 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+/**
+ * @file PerformanceModel.cpp
+ * Performance model.
+ */
+
+#include <geo/geo.h>
+#include <px4_platform_common/events.h>
+#include "PerformanceModel.hpp"
+#include <lib/atmosphere/atmosphere.h>
+
+using namespace atmosphere;
+
+
+// [.] minimum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMinWeightRatio = 0.5f;
+
+// [.] maximum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMaxWeightRatio = 2.0f;
+
+// [m/s] climbrate defining the service ceiling, used to compensate max climbrate based on air density
+static constexpr float kClimbRateAtServiceCeiling = 0.5f;
+
+PerformanceModel::PerformanceModel(): ModuleParams(nullptr)
+{
+	updateParams();
+}
+float PerformanceModel::getWeightRatio() const
+{
+	float weight_factor = 1.0f;
+
+	if (_param_weight_base.get() > FLT_EPSILON && _param_weight_gross.get() > FLT_EPSILON) {
+		weight_factor = math::constrain(_param_weight_gross.get() / _param_weight_base.get(), kMinWeightRatio,
+						kMaxWeightRatio);
+	}
+
+	return weight_factor;
+}
+float PerformanceModel::getMaximumClimbRate(float air_density) const
+{
+	air_density = sanitiseAirDensity(air_density);
+	float climbrate_max = _param_fw_t_clmb_max.get();
+
+	const float service_ceiling = _param_service_ceiling.get();
+
+	if (service_ceiling > 0.0f) {
+		const float ceiling_pressure = getPressureFromAltitude(service_ceiling);
+		const float ceiling_density = getDensityFromPressureAndTemp(ceiling_pressure,
+					      getStandardTemperatureAtAltitude(service_ceiling));
+		const float climbrate_gradient = math::max((_param_fw_t_clmb_max.get() - kClimbRateAtServiceCeiling) /
+						 (kAirDensitySeaLevelStandardAtmos -
+						  ceiling_density), 0.0f);
+		const float delta_rho = air_density - kAirDensitySeaLevelStandardAtmos;
+		climbrate_max = math::constrain(_param_fw_t_clmb_max.get() + climbrate_gradient * delta_rho, kClimbRateAtServiceCeiling,
+						_param_fw_t_clmb_max.get());
+	}
+
+	return climbrate_max / getWeightRatio();
+}
+float PerformanceModel::getTrimThrottle(float throttle_min, float throttle_max, float airspeed_sp,
+					float air_density) const
+{
+	const float throttle_trim = getTrimThrottleForAirspeed(airspeed_sp) * getAirDensityThrottleScale(
+					    air_density) * getWeightThrottleScale();
```

Comment:
so we'd want to delete the weight compensation here, and rather adapt the airspeed setpoint in a separate function -- e.g. it is already applied in the `getCalibratedTrimAirspeed()` method - maybe we should take it out of the nominal trim airspeed getter and consolidate all scale ups in a `compensateCalibratedAirspeed()`? where weight ratio and load factor can both be applied? just to try and avoid doubling calculations accidentally

e.g.

Reply:
@tstastny Aren't those two separate issues? If a vehicle flies at a specifc calibrated airspeed setpoint and now you start to increase weight, then the trim throttle will increase as well, independent from airspeed.

Reply:
Agree, unless induced drag is neglectable.